python/raf/_tvm_op/reduce.py

```diff
@@ -61,31 +67,306 @@ def schedule_sum(attrs, outs, target):
+                        )
+                        remain_total_size //= cand_sizes[idx]
+
+                    # If the product of candidate dividable (block * thread) is too small,
+                    # then the performance may be worse even half2 is enabled. Note that 0.7
+                    # is just a heuristic ratio and may not be optimal for all workloads.
+                    if np.prod(cand_sizes) / (max_block * num_thread) >= 0.7:
+                        num_thread, max_block = cand_sizes
+
+                need_block_split = const_size > max_block * num_thread * vector_width
+            except ValueError:
+                need_block_split = False
+                const_size = 0
+
+            if vector_width > 1:
+                fused, vec = sch[out].split(fused, vector_width)
+                sch[out].vectorize(vec)
+
+            if need_block_split:
+                xo, xi = sch[out].split(fused, factor=num_thread * max_block)
+                bx, tx = sch[out].split(xi, factor=num_thread)
+                sch[out].reorder(bx, tx, xo)
+                sch[out].bind(bx, _tvm.te.thread_axis("blockIdx.x"))
+                sch[out].bind(tx, _tvm.te.thread_axis("threadIdx.x"))
+            else:
+                if const_size != 0 and const_size < num_thread:
+                    bx, tx = sch[out].split(fused, factor=const_size)
+                else:
+                    bx, tx = sch[out].split(fused, factor=num_thread)
+                sch[out].bind(tx, _tvm.te.thread_axis("threadIdx.x"))
+                sch[out].bind(bx, _tvm.te.thread_axis("blockIdx.x"))
+    return sch
+
+
 @schedule_sum.register(["cuda", "gpu"])
 def schedule_sum_cuda(attrs, outs, target):
     # pylint: disable=unused-argument
-    # pylint: disable=invalid-name
     def get_num_elements(axes):
         extents = [int(iv.dom.extent) for iv in axes]
         n_elems = 1
         for extent in extents:
             n_elems *= extent
         return n_elems
 
+    def get_sum_input(tensor):
+        operator = tensor.op
+        if len(operator.input_tensors) != 1:
+            return None
+
+        input_tensor = operator.input_tensors[0]
+        if isinstance(input_tensor.op, _tvm.te.PlaceholderOp):
+            return input_tensor
+        return None
+
     with target:
         out = outs[0]
         num_out_elements = get_num_elements(out.op.axis)
         num_reduce_elements = get_num_elements(out.op.reduce_axis)
 
-        # We want to saturate the GPU cores by parallelization. There are 2 scenarios
-        # 1) Reduce dimension is small - In this case, each thread is responsible for reduction.
-        # The parallelization is across the output elements.
-        # 2) Reduce dimension is large - We want to parallelize the reduction to keep the GPU busy.
-        # Here we fall back to TVM schedule.
+        # Whether the last axis is reduced. Note that axis=-1 should already be proceed in advance.
```

Comment:
What does this comment "should alraedy be proceed" mean?

Reply:
It means axis=-1 should already be changed to axis=ndim-1 before entering this function.

Reply:
got it. the rest of code LGTM. approved.